text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,639 @@
+* Settlement operations are triggered manually by the user.
+* In callback scenarios settlement can be automatically triggered by messaging SDKs
+  based on return values of callbacks.
+
+A "Settle" span should be created for every settlement operation, no matter
+which party triggered it.  SDKs will, in some cases, auto-settle messages in
+push-scenarios when messages are delivered via callbacks. In cases where it is
+possible, it is recommended to create the "Settle" span as a child of the
+"Deliver" span.
+
+"Settle" spans SHOULD link to "Create" or "Publish" spans of the messages that are
+settled, when possible.
+
+## Proposed changes and additions to the messaging semantic conventions
+
+### Operation name
+
+The following operations related to messages are covered by these semantic
+conventions:
+
+| Operation name | Description |
+|----------------|-------------|
+| `publish`      | One or more messages are provided for publishing to an intermediary. |
+| `create`       | A message is created. |
+| `receive`      | One or more messages are requested by a consumer. |
+| `deliver`      | One or more messages are passed to a consumer. |
+| `settle`       | One or more messages are settled. |
+
+For further details about each of those operations refer to the [section about trace structure](#trace-structure).
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+
+The kind of `settle` spans should be set according to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind),
+e. g. it should be set to `CLIENT` if the `settle` spans models a synchronous call
+to the intermediary.
+
+Setting span kinds according to this table ensures that span links between
+consumers and producers always go from a `PRODUCER` span on the producer side to
+a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
+linked traces without the need for additional semantic hints.
+
+### Span relationships
+
+#### Producer spans
+
+"Publish" spans SHOULD be created for operations of providing messages for
+sending or publishing to an intermediary. A single "Publish" span can account
+for a single message, or for multiple messages (in the case of providing
+messages in batches). "Create" spans MAY be created. A single "Create" span
+SHOULD account only for a single message. "Create" spans SHOULD either be
+children or links of the related "Publish" span.
+
+If a "Create" span exists for a message, its context SHOULD be injected into
```

Comment:
I'd like to avoid any normative `MUST` wording regarding span structure and context injection. I can think of cases where  attaching certain contexts is not technically possible or feasible, and I think it makes sense in those cases to allow best-effort instrumentations to be compliant.